### ISSUE TITLE
Middle Langevin Integrator with HMR (3 amu)

### ISSUE THREAD

**Issue by DanielWicz** [openmm/openmm#3117]
I'd like to get better performance (ns/day) and I don't know if it's reasonable to use Middle Langevin Integrator  with HMR.
I've read #2520, nevertheless I can't make a conclusion if it will make reasonable results. Anybody has any experience with such setting ?

**Comment by jchodera** [comment]
There should be no issues running the Middle Langevin integrator (a leapfrog form of [BAOAB](https://royalsocietypublishing.org/doi/full/10.1098/rspa.2016.0138)) with HMR. We've found that [these integrators introduce very little error into configuration space sampled densities](https://www.choderalab.org/publications/2018/2/16/quantifying-configuration-sampling-error-in-langevin-simulations-of-complex-molecular-systems?rq=langevin)) even up to the timestep where they break. So you should be able to safely tune the timestep and hydrogen mass to improve performance.

**Comment by DanielWicz** [comment]
> There should be no issues running the Middle Langevin integrator (a leapfrog form of [BAOAB](https://royalsocietypublishing.org/doi/full/10.1098/rspa.2016.0138)) with HMR. We've found that [these integrators introduce very little error into configuration space sampled densities](https://www.choderalab.org/publications/2018/2/16/quantifying-configuration-sampling-error-in-langevin-simulations-of-complex-molecular-systems?rq=langevin)) even up to the timestep where they break. So you should be able to safely tune the timestep and hydrogen mass to improve performance.

Thanks for quick reply. At the moment I see HMR with MLI (Middle Langevin Integrator) allows for a stable 5 fs timestep in a protein membrane system.
PS. Is the HMR in OpenMM applied to the all non-water molecules or also to the waters ?

**Comment by peastman** [comment]
For what it's worth, the default settings in OpenMM-Setup are a 4 fs timestep and 1.5 amu hydrogen mass.  Those settings seem to give stable integration in a wide range of systems with minimal perturbation of the dynamics.

In the current release, HMR is applied to all hydrogens.  #2975 modifies it to not apply to rigid waters.  That will be in the next release.

**Comment by DanielWicz** [comment]
> For what it's worth, the default settings in OpenMM-Setup are a 4 fs timestep and 1.5 amu hydrogen mass. Those settings seem to give stable integration in a wide range of systems with minimal perturbation of the dynamics.
> 
> In the current release, HMR is applied to all hydrogens. #2975 modifies it to not apply to rigid waters. That will be in the next release.

You know what's ETA for the release with #2975 ? I think HMR on water will affects it's viscosity. And it may affect conformational sampling of protein conformations or properties of the lipid membrane.

**Comment by peastman** [comment]
We haven't scheduled it yet.  I would guess a few months.

**Comment by jchodera** [comment]
@peastman : That reminds me that we wanted to use the [milestone-tagging feature](https://docs.github.com/en/github/managing-your-work-on-github/associating-milestones-with-issues-and-pull-requests) so we can give our users better clarity into which releases various issues/features are expected, and potentially target dates for releases.

**Comment by DanielWicz** [comment]
> There should be no issues running the Middle Langevin integrator (a leapfrog form of [BAOAB](https://royalsocietypublishing.org/doi/full/10.1098/rspa.2016.0138)) with HMR. We've found that [these integrators introduce very little error into configuration space sampled densities](https://www.choderalab.org/publications/2018/2/16/quantifying-configuration-sampling-error-in-langevin-simulations-of-complex-molecular-systems?rq=langevin)) even up to the timestep where they break. So you should be able to safely tune the timestep and hydrogen mass to improve performance.

 I see that I can go up to 5.5fs with mixed precision and HMR = 2 amu, and it crashes when I use single precision. Nevertheless, I'm worried that my results are not realistic or the simulation can be somehow flawed by using such a high time step. There is any way to assess if the results are not somehow "non-realistic" ?

**Comment by rpavlovicz** [comment]
> For what it's worth, the default settings in OpenMM-Setup are a 4 fs timestep and 1.5 amu hydrogen mass. Those settings seem to give stable integration in a wide range of systems with minimal perturbation of the dynamics.

Hi  @peastman, I've been reading some of these exchanges, including #2520 where the current integrator defaults seem to have been settled upon. My question is regarding the 1.5 amu for H that you mention as a default setting. Using OpenMM-Setup, I don't see where this is taking place, at least as a default. 

The example scripts in section 3 of the online manual use a 4fs timestep and constraints=HBonds with the LangevinMiddleIntegrator, but i don't see anything about `hydrogenMass = 1.5*amu`.

Can you help clear up if indeed `hydrogenMass = 1.5*amu` should be used to get the reliable configuration space properties with the 4fs timestep?

**Comment by peastman** [comment]
Yes, that's the default in OpenMM-Setup.  Here are the lines that set the default options for it:

https://github.com/openmm/openmm-setup/blob/931bddaf9fa185e44c79af2c638c22eed3cbe7d4/openmmsetup/openmmsetup.py#L422-L423

**Comment by rpavlovicz** [comment]
Thanks for pointing me to the default options, @peastman. 

This is deviating from the issue topic, but is there a way to set all parameters to default in OpenMM-Setup? My first time using it, all fields were empty and subsequent executions of the code seem to auto-populate all fields with the previously defined parameters.

**Comment by peastman** [comment]
That sounds like something went wrong.  Check the console window you launched it from.  Did any error messages get printed there?

**Comment by rpavlovicz** [comment]
None of the console messages seem like errors to me:

``` * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
127.0.0.1 - - [16/Jun/2021 11:24:54] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [16/Jun/2021 11:24:56] "GET /selectFiles?type=pdb HTTP/1.1" 200 -
127.0.0.1 - - [16/Jun/2021 11:25:08] "POST /configureFiles HTTP/1.1" 200 -
127.0.0.1 - - [16/Jun/2021 11:25:08] "GET /getCurrentStructure HTTP/1.1" 200 -
127.0.0.1 - - [16/Jun/2021 11:25:12] "POST /selectChains HTTP/1.1" 200 -
127.0.0.1 - - [16/Jun/2021 11:25:39] "POST /addHydrogens HTTP/1.1" 200 -
127.0.0.1 - - [16/Jun/2021 11:25:39] "POST /setSimulationOptions HTTP/1.1" 200 -
127.0.0.1 - - [16/Jun/2021 11:26:01] "GET /downloadPackage HTTP/1.1" 200 -
127.0.0.1 - - [16/Jun/2021 11:26:21] "GET /headerControls?quit=Quit HTTP/1.1" 200 -```

The behavior is similar every time I load OpenMM-Setup, with the previous session's fields remaining. Is everything supposed to get set to default with each new launch? Or should be an option somewhere to reset to default?

**Comment by peastman** [comment]
When you say "load", do you just mean reload the page in your web browser?  That should keep everything as before.  If you click the "start over" button, or the "quit" button and then relaunch the application from the console, it should return to defaults.

**Comment by rpavlovicz** [comment]
I meant I was re-launching `openmm-setup` from command line. Everytime I press `quit` and relaunch from command line, the same user inputs are pre-loaded for all newly loaded pdbs. 🤷‍♂️

I personally like this behavior, since I would likely choose to run with the same parameters for all my systems. Except it would be nice to be able to have the option to restore to defaults / recommended settings.

**Comment by peastman** [comment]
Is it possible your web browser is doing it?  Maybe an extension you have installed for remembering forms?
